src/BucketBrigadeCCZ.qs

```diff
@@ -0,0 +1,192 @@
+﻿namespace Qram{
+    
+    open Microsoft.Quantum.Math;
+    open Microsoft.Quantum.Convert;
+    open Microsoft.Quantum.Arrays;
+    open Microsoft.Quantum.Canon;
+    open Microsoft.Quantum.Intrinsic;
+    open Microsoft.Quantum.Arithmetic;
+    open Microsoft.Quantum.Measurement;
+    open Microsoft.Quantum.Diagnostics;
+
+///////////////////////////////////////////////////////////////////////////
+// PUBLIC API
+///////////////////////////////////////////////////////////////////////////
+
+    /// # Summary
+    /// Creates a QRAM type corresponding to a Bucket Brigade scheme that
+    /// heavily parallelizes the T gates by converting Toffolis to CCZs.
+    /// # Input
+    /// ## dataValues
+    /// The data to be stored in the memory.
+    /// ## memoryRegister
+    /// The register that you want to be initialized with the provided data.
+    /// # Output
+    /// An instance of the QRAM type that will allow you to use the memory.
+    operation BucketBrigadeCCZQRAMOracle(dataValues : MemoryCell[], memoryRegister : MemoryRegister) : QRAM {
+        let bank = GeneratedMemoryBank(dataValues);
+
+        for (cell in bank::DataSet) {
+            BucketBrigadeWrite(memoryRegister, cell);
+        }
+
+        return Default<QRAM>()
+            w/ Read <-  BucketBrigadeCCZRead(_, _, _)
+            w/ Write <- BucketBrigadeCCZWrite(_, _)
+            w/ AddressSize <- bank::AddressSize
+            w/ DataSize <- bank::DataSize;
+    }
+
+///////////////////////////////////////////////////////////////////////////
+// INTERNAL IMPLEMENTATION
+///////////////////////////////////////////////////////////////////////////
+    
+    /// # Summary
+    /// Writes a single bit of data to the memory.
+    /// # Input
+    /// ## memoryRegister
+    /// Register that represents the memory you are writing to.
+    /// ## dataValue
+    /// The tuple of (address, data) that you want written to the memory.
+    operation BucketBrigadeCCZWrite(
```

Comment:
Shouldn't these be internal ?

Reply:
Oh nice catch! I'm not sure - some of the stuff I needed to make not internal in order to use them in the tests. Is there a way to like make them available for tests only but not for a user?

Reply:
In C# and VB there is normally the attribute internalsvisibleto one can put one the assembly for that (makes the internals usable for a test library). Don't know if Q# supports that. @crazy4pi314 do you know ?

Reply:
Not that I know of, probably best we try and follow the style guide on where to use? https://docs.microsoft.com/en-us/quantum/resources/contributing/style-guide?tabs=guidance#private-or-internal-names

Reply:
We may have to leave them exposed for now so tests pass, Imma add a big TODO to make this internal once we figure out how to solve the testing thing.